File: algorithms/cracking_the_coding_interview/stacks_balanced_brackets.php
<?php
declare(strict_types=1);

const YES = 'YES';
const NO = 'NO';

/**
 * @param string $brackets
 * @return string
 */
function areBalancedBrackets(string $brackets): string
{
    $stack = [];
    $length = strlen($brackets);

    for ($i = 0; $i < $length; $i++) {
        $bracket = $brackets[$i];
        if ($bracket === '(' || $bracket === '[' || $bracket === '{') {
            array_unshift($stack, $bracket);
            continue;
        }

        if (empty($stack) && ($bracket === ')' || $bracket === ']' || $bracket === '}')) {
            return NO;
        }

        $popFromStack = array_shift($stack);

        if ($bracket === ')' && $popFromStack !== '(') {
            return NO;
        }
        if ($bracket === ']' && $popFromStack !== '[') {
            return NO;
        }
        if ($bracket === '}' && $popFromStack !== '{') {
            return NO;
        }
    }

    if (!empty($stack)) {
        return NO;
    }

    return YES;
}

$handle = fopen('php://stdin', 'r');
fscanf($handle, '%d', $t);

$result = [];
for ($a0 = 0; $a0 < $t; $a0++) {
    fscanf($handle, '%s', $expression);
    $result[] = areBalancedBrackets($expression);
}
echo implode("\n", $result);

File: algorithms/cracking_the_coding_interview/stacks_balanced_brackets.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const (
	YES = "YES"
	NO  = "NO"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	reader.ReadString('\n')

	for {
		brackets, err := reader.ReadString('\n')
		brackets = strings.TrimSpace(brackets)

		fmt.Println(areBalancedBrackets(brackets))

		if err != nil {
			break
		}
	}
}

func areBalancedBrackets(brackets string) string {
	var stack []string

	for _, v := range brackets {
		bracket := string(v)
		if bracket == "(" || bracket == "{" || bracket == "[" {
			stack = append([]string{bracket}, stack...)
			continue
		}

		if len(stack) == 0 {
			return NO
		}

		popFromStack := stack[0]
		stack = stack[1:]

		if bracket == ")" && popFromStack != "(" {
			return NO
		}
		if bracket == "}" && popFromStack != "{" {
			return NO
		}
		if bracket == "]" && popFromStack != "[" {
			return NO
		}
	}

	if len(stack) > 0 {
		return NO
	}

	return YES
}
